Islandora-Devops/isle-dc
# Set Memory limits by container

In the event of a memory leak this should prevent the server from taking over the host and crashing. It also should improve the host environment during build time, improving the stability and expanding the functionality of this build.

I have recently have seen blazegraph and cantaloupe complete consume all my RAM during the build process ultimately crashing docker and other apps. This may be due to something I was doing within the containers but it brought the idea to my attention that a memory leak could cripple a host machine if there is no limits set.

Running `docker stats` should output this. The focus is on **MEM USAGE / LIMIT** and **MEM %**.

```table
CONTAINER ID   NAME                   CPU %     MEM USAGE / LIMIT   MEM %     NET I/O           BLOCK I/O         PIDS
a17f66002bfb   isle-dc_fits_1         0.90%     1.51GiB / 5GiB      30.19%    17kB / 0B         11.7MB / 15.4MB   50
c32664f04b09   isle-dc_activemq_1     0.64%     228.9MiB / 2GiB     11.17%    1.11MB / 1.07MB   3.49MB / 24.4MB   70
cbb19ba8435f   isle-dc_mariadb_1      2.20%     172.1MiB / 1GiB     16.81%    69.7MB / 804MB    9.72MB / 226MB    37
fa149fe62de3   isle-dc_houdini_1      0.55%     15.96MiB / 1GiB     1.56%     16.9kB / 0B       12.3kB / 13.9MB   17
ea787d3cf937   isle-dc_solr_1         0.55%     217.8MiB / 8GiB     2.66%     297kB / 23.8kB    76.8MB / 15.7MB   45
eaa3a58f1f38   isle-dc_hypercube_1    0.59%     15.98MiB / 512MiB   3.12%     16.1kB / 0B       36.9kB / 13.8MB   17
6fe92b3432ab   isle-dc_recast_1       0.56%     16.04MiB / 1GiB     1.57%     15.5kB / 0B       270kB / 13.9MB    17
777517051fbe   isle-dc_homarus_1      0.64%     16.01MiB / 1GiB     1.56%     15.5kB / 0B       123kB / 13.8MB    17
8b3ddc4d8dfa   isle-dc_alpaca_1       1.15%     357.8MiB / 2GiB     17.47%    3.19MB / 2.4MB    9.85MB / 16.7MB   104
994f27f972a5   isle-dc_watchtower_1   2.48%     10.61MiB / 2GiB     0.52%     14.9kB / 0B       18.2MB / 0B       13
5d2daeb8738f   isle-dc_cantaloupe_1   1.32%     732MiB / 5GiB       14.30%    14.4kB / 0B       6.81MB / 17.5MB   54
d082989faf8f   isle-dc_blazegraph_1   0.77%     1.202GiB / 5GiB     24.04%    984kB / 199kB     19.4MB / 494MB    71
20a73f7bc5d5   isle-dc_crayfits_1     0.56%     16.24MiB / 1GiB     1.59%     14.2kB / 0B       36.9kB / 13.8MB   17
8f97c66603e7   isle-dc_milliner_1     4.73%     28.03MiB / 1GiB     2.74%     6.22MB / 2.89MB   8.56MB / 13.9MB   18
35b0146533e1   isle-dc_matomo_1       0.60%     19.37MiB / 1GiB     1.89%     54.1kB / 129kB    147kB / 158MB     17
cb76f0b47c96   traefik                1.25%     19.19MiB / 8GiB     0.23%     6.95MB / 9.66MB   35.3MB / 0B       14
e9bdb01aca57   isle-dc_drupal_1       3.79%     103.3MiB / 5GiB     2.02%     814MB / 76.2MB    88.6MB / 75.7MB   18
4bd3035f6821   isle-dc_fcrepo_1       25.69%    850.7MiB / 5GiB     16.61%    5.07MB / 7.07MB   16.4MB / 20.9MB   77
```

### What changed
Within the **sample.env** file is a variable for each one of these containers using the naming convention _NAME_**_MEMORY_LIMIT**. 
Within each of the containers that a value can be set, a memory limit variable is added with a fallback value reflecting the software vendor's minimal requirements. This "fallback" value is for existing installations that don't already have the _NAME_**_MEMORY_LIMIT** variables within their **.env** file. This anticipates that possibility and doesn't prevent the build process from succeeding.

### How to test
- Before switching to this branch, run `docker stats` on isle-dc to see the normal operation.
- All of the MEM USAGE / **LIMIT** should have a LIMIT that matches your host machine's RAM.
- After switching to this pull request branch, run and either `make up`, `docker-compose up -d` or how ever you plan on testing this then run `docker stats` again and it should have limits that reflect the **sample.env**. 
- If you didn't build from scratch the **LIMIT** values should match those in the docker-compose.xxx.yml files. These are slightly different. The values in the YML files reflect the specified **minimal** requirements from the vendors. The _sample.env_ values are slightly higher to improve build times (ideal configurations).
